sonic_platform_base/sonic_xcvr/api/public/cmis.py

```diff
@@ -1745,4 +1765,246 @@ def get_transceiver_loopback(self):
+        for lane in range(self.NUM_CHANNELS):
+            if (1 << lane) & host_lanemask == 0:
+                continue
+            if cmis_major >= 4: # CMIS v4 onwards
+                data |= (1 << lane)
+            else:               # CMIS v3
+                data &= ~(1 << lane)
+        self.xcvr_eeprom.write(consts.DATAPATH_DEINIT_FIELD, data)
+
+    def get_host_speed(self, ifname):
+        """
+        Get the port speed from the host interface name
+
+        Args:
+            ifname: String, host interface name
+
+        Returns:
+            Integer, the port speed if success otherwise 0
+        """
+        # see HOST_ELECTRICAL_INTERFACE of sff8024.py
+        speed = 0
+        if '400G' in ifname:
+            speed = 400000
+        elif '200G' in ifname:
+            speed = 200000
+        elif '100G' in ifname or 'CAUI-4' in ifname:
+            speed = 100000
+        elif '50G' in ifname or 'LAUI-2' in ifname:
+            speed = 50000
+        elif '40G' in ifname or 'XLAUI' in ifname or 'XLPPI' in ifname:
+            speed = 40000
+        elif '25G' in ifname:
+            speed = 25000
+        elif '10G' in ifname or 'SFI' in ifname or 'XFI' in ifname:
+            speed = 10000
+        elif '1000BASE' in ifname:
+            speed = 1000
+        return speed
+
+    def get_cmis_state(self):
+        """
+        Get the CMIS states
+
+        Returns:
+            Dictionary, the states of module, config error and datapath
+        """
+        state = {
+            'module_state': self.get_module_state(),
+            'config_state': self.get_config_datapath_hostlane_status(),
+            'datapath_state': self.get_datapath_state()
+        }
+        return state
+
+    def get_cmis_application_selected(self, host_lane):
+        """
+        Get the CMIS selected application code of a host lane
+
+        Args:
+            host_lane:
+                Integer, the lane id on the host/system side
+
+        Returns:
+            Integer, the transceiver-specific application code
+        """
+        ap_code = 0
+        if host_lane in range(self.NUM_CHANNELS) and not self.is_flat_memory():
+            field = "{}_{}_{}".format(consts.STAGED_CTRL_APSEL_FIELD, 0, host_lane + 1)
+            ap_code = self.xcvr_eeprom.read(field) >> 4
+
+        return (ap_code & 0xf)
+
+    def get_cmis_application_matched(self, host_speed, host_lanemask):
+        """
+        Get the CMIS application code that matches the specified host side configurations
+
+        Args:
+            host_speed:
+                Integer, the port speed of the host interface
+            host_lanemask:
+                Integer, a bitmask of the lanes on the host side
+                e.g. 0x5 for lane 0 and lane 2.
+
+        Returns:
+            Integer, the transceiver-specific application code
+        """
+        if host_speed == 0 or host_lanemask == 0:
+            return 0
+
+        host_lane_count = 0
```

Comment:
I believe that we have already have a API to get host lane count. can you please check?.

Reply:
Done